src/common/components/giftViews/giftSearchView/giftSearchView.component.js

```diff
@@ -0,0 +1,104 @@
+import angular from 'angular';
+import map from 'lodash/map';
+import forEach from 'lodash/forEach';
+import find from 'lodash/find';
+import filter from 'lodash/filter';
+import {Observable, Subject} from 'rxjs';
+import 'rxjs/add/observable/empty';
+import 'rxjs/add/operator/debounceTime';
+import 'rxjs/add/operator/distinctUntilChanged';
+import 'rxjs/add/operator/do';
+import 'rxjs/add/operator/switchMap';
+
+import giftListItem from 'common/components/giftViews/giftListItem/giftListItem.component';
+import loadingComponent from 'common/components/loading/loading.component';
+
+import designationsService from 'common/services/api/designations.service';
+
+import template from './giftSearchView.tpl';
+
+let componentName = 'giftSearchView';
+
+class GiftSearchViewController {
+
+  /* @ngInject */
+  constructor($log, $scope, designationsService) {
+    this.$log = $log;
+    this.$scope = $scope;
+    this.designationsService = designationsService;
+
+    this.searchSubject = new Subject();
+    this.searchState = 'initial';
+  }
+
+  $onInit(){
+    this.searchHandler();
+  }
+
+  searchHandler(){
+    this.searchSubject
+      .debounceTime(600)
+      .distinctUntilChanged()
+      .do(() => {
+        this.searchState = 'searching';
+      })
+      .switchMap(query => {
+        // switchMap cancels the existing observable when a new event arrives and uses the latest
+        if(query){
+          return this.designationsService.productSearch( {keyword: query} );
+        }else{
+          this.$scope.$apply(() => this.searchState = 'initial');
```

Comment:
Changing the state here wasn't changing the template until something else caused a digest loop. If you have any ideas on how to avoid the apply that would be great. I guess this could be getting executed outside of what angular is watching but that doesn't quite make sense.